grep-0018-modernize-c++.md

```diff
@@ -0,0 +1,417 @@
+./gnuradio-runtime/lib/controlport/thrift/thrift-codebase-shutdown-patch.diff:+    serverTransport_->interrupt();
+./gnuradio-runtime/lib/controlport/thrift/thrift-codebase-shutdown-patch.diff:-    serverTransport_->interrupt();
+```
+
+There are 26 `condition_variable`s, and possibly other trigger points
+that matter.
+
+The work of changing the scheduler and the method of thread
+cancellation is big enough to be out of scope for this GREP, and will
+require a separate one.
+
+Recommendation:
+* Leave these for a future GREP
+
+### `volk_malloc`
+
+Likely all of these calls can be replaced by `volk::vector`, reducing
+manual memory management, improving readability, and reducing chances
+for mistakes like copy-paste ` * sizeof(float)` where ` *
+sizeof(gr_complex)` was intended.
+
+Recommendation:
+* Replace all feasible uses of `volk_malloc` et al with `volk::vector`
+
+Risks: master branch already depends on `volk::vector`, so none.
+
+### `xxx_cast<>`
+
+`static_cast<>` and friends are safer than C style casts, but more
+verbose.
+
+A typical example is:
+
+```c++
+[…]
+  gr_vector_const_void_star& input_items,
+[…]
+  float* in = (float*)input_items[0];
+```
+
+which should be changed to use a `static_cast<>` in order to not cast
+away the constness.
+
+Recommendation:
+* Change them where they make sense, but don't have as a goal to
+  eliminate C style casts where they seem right
+
+Risks: none
+
+### `static_assert`
+
+There are many uses of `assert()` (runtime check) that should be
+`static_assert` (compile time check).
+
+Example:
+
+```c++
+assert(sizeof(fftwf_complex) == sizeof(gr_complex))
+```
+
+Risks: none
+
+### Default member initialization
+
+Pre-C++11 there were only two ways to initialize values, both in the
+constructor:
+
+```c++
+class obj {
+  int d_a, d_b, d_c;
+  std::string d_s;
+public:
+  obj(int a)
+    : d_a(a),
+      d_b(0),
+      d_s()   // Completely unnecessary
+  {
+      d_c = 3;
+  }
+  obj(int a, int b)
+    : d_a(a), d_b(b)  // Repeated init of d_a.
+  {
+      d_c = 3; // Repeated init of d_c;
+  }
+};
+```
+
+This causes code duplication.
+
+In C++11 default member values can be set at declaration time
+(normally in header files):
+
+```c++
+class obj {
+  int d_a = 0;   // Or int d_a{0};
+  int d_b = 0;
+  int d_c = 3;
+public:
+  obj(int a)
+    : d_a(a)
+  {
+  }
+  obj(int a, int b):
+    : d_a(a), d_b(b)    // Duplication of d_a(a).
+  {
+  }
+};
+```
+
+To completely avoid duplication one constructor can call another:
+
+```c++
+class obj {
+  int d_a = 0;   // Or int d_a{0};
+  int d_b = 0;
+  int d_c = 3;
+public:
+  obj(int a)
+    : d_a(a)
+  {
+  }
+  obj(int a, int b)
+    : obj(a) // Call the other constructor.
+  {
+    d_b = b; // Can't be put in construct initializer when
+             // construction delegation is used.
+  }
+};
+```
+
+Recommendation:
+
+* Don't use constructor delegation with a non-empty body. It's
```

Comment:
Again, it's not clear if you're advocating for making that a coding guideline, or if you want to change existing code.

Reply:
Most of this is C++11, so to backwards change needed.

But I'm elaborating